python/dgl/nn/utils.py

```diff
@@ -0,0 +1,80 @@
+"""Graph construction utilities used in dgl.nn modules"""
+# pylint: disable= no-member, arguments-differ
+
+import numpy as np
+
+from ..graph import DGLGraph
+from ..batched_graph import BatchedDGLGraph, batch
+from ..graph_index import from_csr
+from .. import backend as F
+
+
+__all__ = ['_create_fully_connected_graph', '_create_bipartite_graph',
+           '_create_graph_from_num_nodes']
+
+def _create_fully_connected_graph(graph):
+    r""" Create a fully connected graph upon the given graph."""
+    if isinstance(graph, BatchedDGLGraph):
+        g_list = []
+        for n_nodes in graph.batch_num_nodes:
+            indptr = np.arange(0, n_nodes * n_nodes + 1, n_nodes)
+            indices = np.tile(np.arange(n_nodes), n_nodes)
+            g_list.append(
+                DGLGraph(from_csr(indptr, indices, False, 'out'), readonly=True)
+            )
+        return batch(g_list)
+    else:
+        n_nodes = graph.number_of_nodes()
+        indptr = np.arange(0, n_nodes * n_nodes + 1, n_nodes)
+        indices = np.tile(np.arange(n_nodes), n_nodes)
+        return DGLGraph(from_csr(indptr, indices, False, 'out'), readonly=True)
+
+def _create_bipartite_graph(graph_u, graph_v):
+    r""" Create a bipartite G(U, V, E) graph with nodes in graph_u as U
+     and nodes in graph_v as V"""
+    if isinstance(graph_u, BatchedDGLGraph):
+        g_list = []
+        if graph_u.batch_size != graph_v.batch_size:
+            raise KeyError('Batch size of graph_u does not equal that of graph_v.')
+        u_list, v_list = [], []
+        v_shift = 0
+        for n_nodes_u, n_nodes_v in zip(graph_u.batch_num_nodes, graph_v.batch_num_nodes):
+            indptr = np.concatenate([
+                np.arange(0, n_nodes_u * n_nodes_v + 1, n_nodes_v),
+                np.full(n_nodes_v, n_nodes_u * n_nodes_v, dtype=np.int64)
+            ])
+            indices = np.tile(np.arange(n_nodes_u, n_nodes_u + n_nodes_v), n_nodes_u)
+            g_list.append(
+                DGLGraph(from_csr(indptr, indices, False, 'out'), readonly=True)
+            )
+            u_list.append(np.arange(v_shift, v_shift + n_nodes_u))
+            v_list.append(np.arange(v_shift + n_nodes_u, v_shift + n_nodes_u + n_nodes_v))
+            v_shift += n_nodes_u + n_nodes_v
+        return batch(g_list),\
+               F.zerocopy_from_numpy(np.concatenate(u_list).astype(np.int64)),\
+               F.zerocopy_from_numpy(np.concatenate(v_list).astype(np.int64))
+    else:
+        n_nodes_u = graph_u.number_of_nodes()
+        n_nodes_v = graph_v.number_of_nodes()
+        u_list, v_list = np.arange(n_nodes_u), np.arange(n_nodes_u, n_nodes_u + n_nodes_v)
+        indptr = np.concatenate([
+            np.arange(0, n_nodes_u * n_nodes_v + 1, n_nodes_v),
+            np.full(n_nodes_v, n_nodes_u * n_nodes_v, )
+        ])
```

Comment:
It seems that this graph is a true bipartite graph that does not have square adjmat. Constructing a DGLGraph from this adjmat is problematic because when calling `number_of_nodes`, it returns `n_nodes_u` instead of `n_nodes_u + n_nodes_v`. This might cause some bugs because `DGLGraph` is not meant for this and we should actually forbidden such behavior. I would suggest
(1) Move the bipartite graph related modules out of this PR, so we could review and merge this.
(2) Wait for bipartite graph implementation which @BarclayII is currently working on.
(3) Finish the rest in another PR.
What do you think?

Reply:
I'll refactor the code where `_create_bipartite_graph` is called and remove these functions.
Why this DGLGraph would return `n_nodes_u` instead of `n_nodes_u + n_nodes_v`, in my experiment it returns `n_nodes_u`.